storage e2e: automate hostpath YAML updates, update sidecars but not driver

pohly opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

We always want to test with the latest plugin and sidecar releases. In this particular case it gets rid of code which still used v1beta1 APIs.

#### Which issue(s) this PR fixes:

Fixes: https://github.com/kubernetes/kubernetes/issues/99957

#### Special notes for your reviewer:

The mirroring of the YAML files was done with a script and should be reproducible and automatic from now on.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
storage e2e: automate hostpath YAML updates, hostpath v1.6.2

Mirroring the various YAML files by hand is tedious. The new
update-hostpath.sh does all the necessary steps automatically.

The result is now a bit more consistent with the upstream repos in the
sense that the original file names and paths for the RBAC YAML files
are used.

The csi-hostpath-testing.yaml is included for the sake of
completeness, but not used during E2E testing.

The new hostpath driver release is v1.6.2, which adds the
external-health-monitor for the first time.

pohly commented:
/retest

pohly commented:
/assign @jsafrane

liggitt commented on test/e2e/testing-manifests/storage-csi/external-provisioner/rbac.yaml:
```diff
@@ -84,6 +93,21 @@ rules:
+- apiGroups: ["storage.k8s.io"]
+  resources: ["csistoragecapacities"]
+  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
+# The GET permissions below are needed for walking up the ownership chain
+# for CSIStorageCapacity. They are sufficient for deployment via
+# StatefulSet (only needs to get Pod) and Deployment (needs to get
+# Pod and then ReplicaSet to find the Deployment).
```
does this mean that pods created from job / cronjob / replicationcontroller / custom-pod-spec-containing objects don't work properly with CSIStorageCapacity?

pohly replied:
The RBAC rules need to be adapted depending on how the external-provisioner gets deployed. This is basically just an example of how to do it, not necessarily what everyone has to use.

liggitt commented on test/e2e/testing-manifests/storage-csi/external-snapshotter/csi-snapshotter/rbac-csi-snapshotter.yaml:
```diff
@@ -23,9 +25,13 @@ rules:
-  - apiGroups: [""]
-    resources: ["secrets"]
-    verbs: ["get", "list"]
+  # Secret permission is optional.
```
happy to see this become optional. would be even happier if the example scoped to a particular namespace/name instead of giving global list permission

pohly replied:
Same here: this is just an example. I suppose we can make the example better by explaining more about the rationale behind the current content, though.

@cc @xing-yang

msau42 replied:
I think we need the secret RBACs for the mock driver secret test: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-gce-storage-snapshot/1384443613999534080/

liggitt replied:
can we narrowly grant to the specific secrets required by the test?

msau42 replied:
Yes we should be able to grant it to the e2e namespace.

pohly replied:
Which E2E namespace? The one of the driver or where test objects are created?

msau42 replied:
It looks like the test namespace: https://github.com/kubernetes/kubernetes/blob/6ede5ca95f78478fa627ecfea8136e0dff34436b/test/e2e/storage/csi_mock_volume.go#L1606

pohly replied:
A Pod running in one namespace cannot be granted access to secrets in one specific other namespace, right? If I remember correctly, the `Role` implicitly is for the namespace in which it gets created, which is the driver namespace.

liggitt replied:
You can grant permissions to a service account in another namespace, see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-binding-examples

The role/rolebinding have to be located in the namespace you want to grant access to, and the serviceaccount subject must be fully qualified:

```
subjects:
- kind: ServiceAccount
  name: some-service-account
  namespace: some-namespace
```

pohly replied:
Re-enabling secret access via ClusterRole was straight-forward (famous last words - the additional commit that I just pushed only passed local build testing). Adding a new service account and associated bindings is going to be more complicated. How important is that?

msau42 replied:
In the interest of time, I would like to keep this PR focused on the immediate issue of updating the sidecars to use v1.CSINode (or have a new PR that only updates the sidecars). Any additional improvements to the tests, let's consider separately.

liggitt replied:
> In the interest of time, I would like to keep this PR focused on the immediate issue of updating the sidecars to use v1.CSINode (or have a new PR that only updates the sidecars). Any additional improvements to the tests, let's consider separately.

sounds good

msau42 replied:
https://github.com/kubernetes/kubernetes/issues/101334

pohly commented on test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-plugin.yaml:
```diff
@@ -34,9 +34,39 @@ spec:
       labels:
         app: csi-hostpathplugin
     spec:
+      serviceAccount: csi-external-health-monitor-controller
```
This can be fixed in an upcoming release of the csi-driver-host-path. Here in k/k the files are supposed to be verbatim copies.

pohly replied:
https://github.com/kubernetes-csi/csi-driver-host-path/pull/271

pohly commented on test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-plugin.yaml:
```diff
@@ -61,15 +91,12 @@ spec:
             - "--endpoint=$(CSI_ENDPOINT)"
             - "--nodeid=$(KUBE_NODE_NAME)"
-            # The only difference to github.com/kubernetes-csi/csi-driver-host-path/deploy
-            # - we have a tests that checks node limits.
```
I need to check that. I ran some tests locally and those worked, but not everything. The fuil test in the CI run showed that further work is needed.

pohly replied:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1376804369592225792 had lots of other, unrelated errors. Must have been a bad host or load. I ran it again, but still no luck.

I checked some of the failed csi-hostpath tests locally and those worked for me. Some tests also passed in the Prow job, so it shouldn't be a general problem with the driver.

Testing on GCE is failing more consistently: the hostpath container doesn't come up.

Hmm, why does it merge YAML manifests as part of the job preparation?! https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-gce-csi-serial/1377185922327515136/build-log.txt

Time for a rebase...

pohly replied:
The merge is probably normal (merge instead of fast forward, with no other changes to those files on master).

pohly replied:
I found the volume limits test and moved the adding of `--maxvolumelimits` parameter to csi.go - that's where the manifests can and should be updated while loading them for tests.

pohly replied:
In principle the updated driver works now. What worries me a bit is the number of tests that time out - not just for storage, but also for other areas (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1377330845706096640). If this is an overall problem, then the PR is ready. But if it is somehow worse with these changes, then we need to investigate further.

/retest

pohly replied:
I had a hunch that the new external-health-monitor sidecars might cause additional overhead, so I disabled them. This is possible because they are not actually needed for any test at the moment.

That made a striking difference:
- without them, pull-kubernetes-e2e-kind finished in 39m with just one unrelated error: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1377513227537092608
- with them, the same job failed both before (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1377369246706700288) and after (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1377531518573875200) the one which succeeded.

That it fails consistently when the sidecars enabled at different times pretty much rules out that the state of the Prow cluster is to blame for the slowdown. Also, other PRs are getting tested (mostly) successfully.

pohly replied:
@xing-yang ^^^

pohly replied:
Or perhaps not. Now e2e-kind also failed horribly without these sidecars: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637

pohly replied:
The problem seems to be elsewhere. The new hostpath driver fails to start up repeatedly while checking mount points:
```
==== start of pod log for container csi-hostpathplugin-0/hostpath ====
I0401 09:26:04.587557       1 hostpath.go:128] Driver: csi-hostpath-ephemeral-6690 
I0401 09:26:04.587617       1 hostpath.go:129] Version: v1.6.2
Failed to initialize driver: failed to get capacity info: no such file or directory
==== end of pod log for container csi-hostpathplugin-0/hostpath ====
```

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1377544760671080448/artifacts/_sig-storage_CSI_Volumes/_Driver_csi-hostpath_/_Testpattern_Generic_Ephemeral-volume_default_fs_immediate-binding_ephemeral/should_support_two_pods_which_share_the_same_volume/csi-hostpathplugin-0-hostpath.log

Unfortunately the error message is not very informative. It comes from here:

https://github.com/kubernetes-csi/csi-driver-host-path/blob/a7442639da30b79a09438e96400bc54602fc068f/pkg/hostpath/hostpath.go#L522-L540

That code is new, it was not yet in the 1.4.0 release that was used before this PR. It was added in https://github.com/kubernetes-csi/csi-driver-host-path/pull/210

@fengzixu any idea why it is failing?

xing-yang replied:
@pohly we have discussed in the CSI meeting that we don't need new e2e tests for health monitoring now because this feature only reports events on PVCs. There's no reliable way of verifying an event is reported as it can disappear.  So until we have figured out a better way, we only test this in unit tests.

pohly replied:
We root-caused this in https://github.com/kubernetes-csi/csi-driver-host-path/pull/210#discussion_r605592438

pohly commented:
/retest

pohly committed:
storage e2e: disable health check containers

They are not needed for any of the tests and in practice apparently
caused enough overhead that even unrelated tests timed out. For
example, in the pull-kubernetes-e2e-kind test, 43 out of 5771 tests
failed, including tests from sig-node, sig-cli, sig-api-machinery,
sig-network.

pohly commented:
/retest

liggitt commented:
what is remaining on this? it is blocking https://github.com/kubernetes/kubernetes/pull/99840

fengzixu commented:
> what is remaining on this? it is blocking #99840, and means master CI will start failing as soon as 1.22 hits 1.22.0-beta.0

Sorry for the inconenience. My work is heavy this week.  The basic implementation of that fixing approach is finished. I am implementing unit test now. Hope I can submit PR during the weekend

@liggitt

msau42 commented:
@pohly @fengzixu as an immediate mitigation since this is blocking the 1.22 v1beta CSINode removal, can we first do a quick fix to update the sidecars but keep the hostpath driver at v1.4.0?

msau42 commented:
My understanding is that:
1. Tests were passing when using v1.4.0 of hostpath driver
2. We need to upgrade the CSI sidecars to stop using v1beta1.CSINode.
3. The newer version of the hostpath driver is causing performance issues.
4. We don't necessarily need to upgrade the hostpath driver version to unblock 2.

pohly committed:
storage e2e: downgrade hostpath driver

This is a temporary workaround until a fixed driver is available.

pohly commented on test/e2e/testing-manifests/storage-csi/update-hostpath.sh:
```diff
@@ -0,0 +1,120 @@
+
+set -xe
+
+# Can be a tag or a branch.
```
Done.

pohly commented:
@msau42 is right, we can upgrade the sidecars while keeping the driver at the working version. I didn't do that because it partly defeats the purpose of this PR (automate the YAML file update), partly because it shouldn't have taken that long to fix the driver.

For the sake of avoiding breakage in Kubernetes, I added one commit with a manual downgrade of the driver to the working version.

pohly commented:
/retest

pohly commented:
Is `pull-kubernetes-e2e-gce-storage-snapshot` not compatible with current snapshotter? It's an optional job, I'm just wondering.

msau42 commented:
> Is `pull-kubernetes-e2e-gce-storage-snapshot` not compatible with current snapshotter? It's an optional job, I'm just wondering.

https://github.com/kubernetes/kubernetes/pull/100637#discussion_r616911550

pohly committed:
storage e2e: patch in RBAC rules for secrets

In one mock test, the snapshotter needs permission to read
secrets. That was disabled in the RBAC files of recent releases. We
need to patch it back in during deployment.

pohly commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

pohly commented:
/retest

Unrelated test failures in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind/1384583976571637760 and https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100637/pull-kubernetes-e2e-kind-ipv6/1384583978249359360 ...

pohly commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

The patch was almost right, but not quite... fixed and now also tested locally.

pohly commented:
/retest

pohly commented:
/assign @msau42 

For review and approval.

msau42 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -220,16 +228,47 @@ func (h *hostpathCSIDriver) PrepareTest(f *framework.Framework) (*storageframewo
+			return err
+		}
+
+		// Remove csi-external-health-monitor-agent and
```
Was this actually the cause? or was it https://github.com/kubernetes-csi/csi-driver-host-path/pull/210#discussion_r605592438?

If we're creating so many sidecars/controllers or they're putting unnecessary load on the system, that's also a problem that we need to investigate (separately).

pohly replied:
To be honest, I don't remember. Let's merge this as-is and then try to enable it again to see whether it makes things worse.

https://github.com/kubernetes/kubernetes/issues/101336

msau42 commented on test/e2e/testing-manifests/storage-csi/hostpath/hostpath/csi-hostpath-testing.yaml:
```diff
@@ -0,0 +1,64 @@
+  selector:
+    app: csi-hostpath-socat
+  ports:
+  - port: 10000 # fixed port inside the pod, dynamically allocated port outside
```
Must we deploy this? We've had issues in the past where other e2e tests have chosen conflicting node ports.

pohly replied:
We don't deploy it (not listed in csi.go). It's here because the script copies the entire directory without any special cases for files that are not needed.

msau42 commented on test/e2e/testing-manifests/storage-csi/update-hostpath.sh:
```diff
@@ -0,0 +1,128 @@
+# Main YAML files.
+mkdir hostpath
+cat >hostpath/README.md <<EOF
+The files in this directory are exact copys of "kubernetes-latest" in
```
typo: copies

pohly replied:
Fixed and pushed.

msau42 commented:
/lgtm
/approve

pohly commented:
We discussed this before merging and the conclusion was that the newer sidecars should not use any beta APIs anymore.

Do you have tests that still fail despite the update?

msau42 commented:
Yes, in the latest round of test failures I verified that volumes were successfully provisioned and mounted. The new failures seem to be in pod sandbox creation.

liggitt commented:
yes, many CSI tests in https://github.com/kubernetes/kubernetes/pull/99840#issuecomment-825163130 (which actually turns off the beta APIs) still fail

BenTheElder commented:
https://github.com/kubernetes/kubernetes/pull/99840#issuecomment-825166188 for future context. thanks!

caesarxuchao commented:
Hi @msau42 @pohly,

Would this PR cause any change to kube-apiserver's egress traffic? Commands like`kubectl exec/attach/proxy/logs` will cause egress traffic of kube-apiserver.

I asked this question because the [ci-kubernetes-e2e-gci-gce-network-proxy-grpc ](https://prow.k8s.io/job-history/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-network-proxy-grpc?buildId=1385142123086483456) tests started to fail after this PR was merged. I couldn't see how this PR would cause the test failures. The only hypothesis I had so far was that perhaps the containers that got changed in this PR were issuing `kubectl exec/attach/proxy/logs` commands. Were they?

pohly commented:
> The only hypothesis I had so far was that perhaps the containers that got changed in this PR were issuing kubectl exec/attach/proxy/logs commands. Were they?

I don't think so.
